File: apple/Sources/ReactNativeWatchOS/Modifiers/Focused.swift
import SwiftUI
import ReactNativeWatchOSCxx

/// SwiftUI `.focused(_:)` — a two-way focus binding. JS owns the focused
/// state via the `value` param; we hold a local `@FocusState` mirror so focus
/// reflects immediately when the system moves it (Digital Crown, taps), fire
/// the JS `handler` with the new boolean on change, then converge back to the
/// authoritative `value` from the next snapshot.
///
/// `@FocusState` / `.focused(_:)` are watchOS 8+, under the watchOS 9
/// deployment target, so no availability gate is needed.
enum RNWFocusedModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("focused") { view, params, ctx in
            AnyView(view.modifier(FocusedMod(
                value: params.bool("value") ?? false,
                handlerId: params.int("handler"),
                bus: ctx.bus
            )))
        }
    }
}

/// Concrete `ViewModifier` so we can hang `@FocusState` off it. The applier
/// re-creates it each render with the latest `value`; `@FocusState` persists
/// because the modified view's identity is stable (`.id(node.tag)` upstream).
private struct FocusedMod: ViewModifier {
    let value: Bool
    let handlerId: Int?
    let bus: RNWEventBus

    @FocusState private var focused: Bool

    func body(content: Content) -> some View {
        content
            .focused($focused)
            // Mirror the initial / updated JS value into the focus state.
            .onAppear { if focused != value { focused = value } }
            .onChange(of: value) { newValue in
                // JS pushed an authoritative value — converge.
                if focused != newValue { focused = newValue }
            }
            .onChange(of: focused) { newFocused in
                // System/user moved focus — notify JS, but only when it
                // diverges from what JS already holds (skip convergence echoes).
                if newFocused != value, let handlerId {
                    bus.fire(handlerId, payload: newFocused)
                }
            }
    }
}

File: src/modifiers/focused.ts
import { createModifier } from './createModifier';

export interface FocusedParams {
  /// JS-owned source of truth for whether this view is focused. The native
  /// side mirrors this into a `@FocusState` and converges on changes.
  value: boolean;
  /// Fired with the new boolean whenever focus enters or leaves the view.
  handler?: (focused: boolean) => void;
}

/// SwiftUI `.focused(_:)` — a two-way focus binding. JS owns the focused
/// state via `value`; the native side holds a `@FocusState` mirror so focus
/// reflects immediately, fires `handler` with the new value on change, and
/// converges back to the JS-supplied `value`. Params `value`, `handler`.
export function focused(params: FocusedParams) {
  return createModifier('focused', params);
}
